fix(tts): add missing supported timeline command to TtsSessionAdapter

ddfreiling opened the pull request:
without this command the session-adapter didn't properly report the current timeline to a media3.MediaSession

ddfreiling committed:
fix(tts): add missing supported timeline command to TtsSessionAdapter

without this command the session-adapter didn't properly report the current timeline to a media3.MediaSession

qnga commented:
What is timeline used for?

ddfreiling commented:
If using the [ForwardingSimpleBasePlayer](https://developer.android.com/reference/androidx/media3/common/ForwardingSimpleBasePlayer) to modify MediaSession, it has a `getState` function to build the player state, which depends on the adapter player allowing the `COMMAND_GET_TIMELINE` to be called, which currently `TtsSessionAdapter` does not.
<img width="848" height="810" alt="Screenshot 2025-10-07 at 13 57 15" src="https://github.com/user-attachments/assets/c9bf036c-ca1d-4fd0-8780-57e48cc4dc2d" />

qnga commented:
I understand that. But what does this getState do better with the timeline? I'm not entirely sure my timeline implementation is good or useful in the adapter, that's why I ask.

ddfreiling commented:
Without a timeline the SimpleBasePlayer dies on an assert that timeline must not be empty if the player is not in IDLE state. For us this crashes the entire MediaSession when we call `openSession` on it with a TtsNavigator.

Exception:
```
SimpleBasePlayer: Empty playlist only allowed in STATE_IDLE or STATE_ENDED
```

As long as SimpleBasePlayer sees a non-empty timeline, everything seems to work.

I guess this only happens, since we wrap the `navigator.asMedia3Player()` in a `ForwardingSimpleBasePlayer` to better control certain media-button events and notification button layout. Maybe there's another way to do this?

qnga commented:
Ok, I see. If you want a specific notification button layout, you might expect a better timeline than my shitty one. Are you expecting resource-based navigation in the notification? I might have not exposed it because of the notification actually, I need to check.

qnga commented:
As far as I understand and remember the missing timeline command is more of a feature than a bug. To prevent your crash you should be able to provide a timeline by yourself by overriding `getState` or something of that ilk. If for some some reason that's not enough, please report your use case in detail.

ddfreiling commented:
@qnga turns out I have to completely duplicate/re-implement the entire `getState` method of ForwardingSimpleBasePlayer. Because if I call super.state and just expand on that, I still get the assert exception:
```
Empty playlist only allowed in STATE_IDLE or STATE_ENDED
```
So basically I have to copy their implementation but leave out this non-empty playlist assert and the check on `AvailableCommands.COMMAND_GET_TIMELINE`.

Maybe the TtsSessionAdapter should just report a single item on the playlist when active?

ddfreiling committed:
Merge remote-tracking branch 'origin' into fix/tts-session-timeline

* origin:
  Update dependencies and JVM target (#717)
  Prevent multiple OutlineFragment instances on repeated menu clicks (#719)

qnga commented:
A single-item timeline sounds reasonable. As mentioned above, my main concern is about the notification. Any solution not making the backward and forward buttons appear in the default notification would be fine with me as this was the intended behavior.

ddfreiling committed:
fix: make TtsSessionAdapter report a single-item timeline

this ensures support for ForwardingSimpleBasePlayer and showing a Notification while playing, but without skip buttons

ddfreiling commented:
What do you think about this solution? In the test-app I'm seeing a simple notification with just a pause/play button now, which also supports wrapping the player in a ForwardingSimpleBasePlayer.

qnga commented:
That looks fine!
